### PR TITLE
Add debug interpreter flag

### PR DESCRIPTION
**Pull request by rsora**
**NOTE:** This PR requires https://github.com/arduino/ArduinoCore-samd/pull/506 to be merged

This PR:
- Removes ugly https://github.com/arduino/arduino-cli/pull/597 hotfix
- Adds `--interpreter` flag and protobuf property to select the desired debug commands interpreter
- Removes mandatory `port` property, not always necessary to use the debugger
- Updates the client example accordingly

### CODE REVIEW

**Comment by rsora** [comment]
cc @spoenemann

**Comment by per1234** [comment]
Should the `interpreter` property be documented in https://arduino.github.io/arduino-cli/platform-specification/?

**Comment by cmaglie** [comment]
> Should the interpreter property be documented in https://arduino.github.io/arduino-cli/platform-specification/?

Yes.

Also the `--interpreter` flag is `gdb`-specific, the real question is: are we going to ever support anything different from a GDB server for the debugger?

If the answer is **no**, as I think, then we should also state explicitly in the docs and in the cli help that the `debug` command will start a **GDB server** session with whatever middleware is needed.

Probably is a good idea to change a bit the recipes in the `platform.txt` from `tools.gdb.debug.pattern` to something like `tools.openocd.debug.pattern` (since we will always use `gdb` and what changes is the middleware).

File: commands/debug/debug.go [hunk 1]
```diff
@@ -227,19 +221,27 @@ func getCommandLine(req *dbg.DebugConfigReq, pm *packagemanager.PackageManager)
...
+		toolProperties.Set("interpreter", "console")
 	}
 
 	// Build recipe for tool
 	recipe := toolProperties.Get("debug.pattern")
-	// REMOVEME: hotfix for samd core 1.8.5
-	if recipe == `"{path}/{cmd}" --interpreter=mi2 -ex "set pagination off" -ex 'target extended-remote | {tools.openocd.path}/{tools.openocd.cmd} -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0"' {build.path}/{build.project_name}.elf` {
-		recipe = `"{path}/{cmd}" --interpreter=mi2 -ex "set remotetimeout 5" -ex "set pagination off" -ex 'target extended-remote | "{tools.openocd.path}/{tools.openocd.cmd}" -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0"' "{build.path}/{build.project_name}.elf"`
-	}
+
```
Comment (cmaglie) [inline commands/debug/debug.go]: Why don't we just update this, so 1.8.5 core will still work until we publish 1.8.6+?

**Comment by cmaglie** [comment]
> from tools.gdb.debug.pattern to something like tools.openocd.debug.pattern

or maybe `tools.gdb-openocd.debug.pattern`?

File: commands/debug/testdata/custom_hardware/arduino-test/samd/platform.txt [hunk 2]
```diff
@@ -228,4 +228,4 @@ tools.gdb.path={runtime.tools.arm-none-eabi-gcc-7-2017q4.path}/bin/
 tools.gdb.cmd=arm-none-eabi-gdb
 tools.gdb.cmd.windows=arm-none-eabi-gdb.exe
 
-tools.gdb.debug.pattern="{path}/{cmd}" -ex 'target extended-remote | {tools.openocd.path}/{tools.openocd.cmd} -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0" -c init -c halt' {build.path}/{build.project_name}.elf
\ No newline at end of file
+tools.gdb.debug.pattern="{path}/{cmd}" --interpreter={interpreter}  -ex 'target extended-remote | {tools.openocd.path}/{tools.openocd.cmd} -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0" -c init -c halt' {build.path}/{build.project_name}.elf
```
Comment (cmaglie) [inline commands/debug/testdata/custom_hardware/arduino-test/samd/platform.txt]: Let's add the default also inside platform.txt with:
```
tools.gdb.interpreter=console
```

**Comment by rsora** [comment]
> If the answer is **no**, as I think, then we should also state explicitly in the docs and in the cli help that the `debug` command will start a **GDB server** session with whatever middleware is needed.

We are already doing it [here](https://github.com/arduino/arduino-cli/blob/97c4f004cc1cdbd5609efeb66568880ebf95f061/cli/debug/debug.go#L46), is that correct?

**Comment by cmaglie** [comment]
> We are already doing it here, is that correct?

Yes

File: commands/debug/testdata/custom_hardware/arduino-test/samd/platform.txt (hunk 2 above)
Comment (rsora) [reply commands/debug/testdata/custom_hardware/arduino-test/samd/platform.txt]: See https://github.com/arduino/ArduinoCore-samd/pull/506/commits/647cd2c2115a2d6cca306b126311023fcc55bf70

File: commands/debug/debug.go [hunk 3]
```diff
@@ -227,19 +221,32 @@ func getCommandLine(req *dbg.DebugConfigReq, pm *packagemanager.PackageManager)
...
+		toolProperties.Set("interpreter", "console")
 	}
 
 	// Build recipe for tool
 	recipe := toolProperties.Get("debug.pattern")
-	// REMOVEME: hotfix for samd core 1.8.5
+
+	// REMOVEME: hotfix for samd core 1.8.5/1.8.6
 	if recipe == `"{path}/{cmd}" --interpreter=mi2 -ex "set pagination off" -ex 'target extended-remote | {tools.openocd.path}/{tools.openocd.cmd} -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0"' {build.path}/{build.project_name}.elf` {
 		recipe = `"{path}/{cmd}" --interpreter=mi2 -ex "set remotetimeout 5" -ex "set pagination off" -ex 'target extended-remote | "{tools.openocd.path}/{tools.openocd.cmd}" -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0"' "{build.path}/{build.project_name}.elf"`
```
Comment (cmaglie) [inline commands/debug/debug.go]:
```suggestion
		recipe = `"{path}/{cmd}" --interpreter={interpreter}  -ex 'target extended-remote | {tools.openocd.path}/{tools.openocd.cmd} -s "{tools.openocd.path}/share/openocd/scripts/" --file "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "gdb_port pipe" -c "telnet_port 0" -c init -c halt' {build.path}/{build.project_name}.elf`
```

Comment (rsora) [reply commands/debug/debug.go]: see https://github.com/arduino/arduino-cli/pull/626/commits/021d460ccc48ec877ad5fa6fa14cda8409054a2c

**Pull request merged**